src/internal/connector/onedrive/restore.go

```diff
@@ -696,26 +797,55 @@ func restoreData(
 		return "", details.ItemInfo{}, err
 	}
 
-	// Get a drive item writer
-	w, err := driveItemWriter(ctx, service, driveID, ptr.Val(newItem.GetId()), ss.Size())
-	if err != nil {
-		return "", details.ItemInfo{}, err
-	}
+	var (
+		w       io.Writer
+		written int64
+	)
 
-	iReader := itemData.ToReader()
-	progReader, closer := observe.ItemProgress(
-		ctx,
-		iReader,
-		observe.ItemRestoreMsg,
-		clues.Hide(name),
-		ss.Size())
+	// This is just to retry file upload, the uploadSession creation is not retried here
+	// We need extra logic to retry file upload as we have to pull the file again from kopia
+	for i := 0; i <= maxUploadRetries; i++ {
+		// Get a drive item writer
+		w, err = driveItemWriter(ctx, service, driveID, ptr.Val(newItem.GetId()), ss.Size())
+		if err != nil {
+			return "", details.ItemInfo{}, err
+		}
```

Comment:
The driveWriter doesn't need to be re-created on retry, and can be moved out of the loop.

Reply:
Graph does not seem to like reusing the endpoint. Let me look into it along with resuming uploads (created an issue https://github.com/alcionai/corso/issues/3521).